lib/charms/tls_certificates_interface/v2/tls_certificates.py

```diff
@@ -1360,12 +1393,70 @@ def _on_relation_broken(self, event: RelationBrokenEvent) -> None:
         Returns:
             None
         """
-        relation = self.model.get_relation(self.relationship_name)
-        if not relation:
+        self.on.all_certificates_invalidated.emit()
+
+    def _on_secret_expired(self, event: SecretExpiredEvent) -> None:
+        """Triggered when a certificate is set to expire.
+
+        Loads the certificate from the secret, and will emit 1 of 2
+        events.
+
+        If the certificate is not yet expired, emits CertificateExpiringEvent
+        and updates the expiry time of the secret to the exact expiry time on
+        the certificate.
+
+        If the certificate is expired, emits CertificateInvalidedEvent and
+        deletes the secret.
+
+        Args:
+            event (SecretExpiredEvent): Juju event
+        """
+        if not event.secret.label or not event.secret.label.startswith(f"{LIBID}-"):
             return
-        if not relation.data[relation.app].get("certificates"):  # type: ignore[index]
+        csr = event.secret.label[len(f"{LIBID}-") :]
+        certificate_dict = self._find_certificate_in_relation_data(csr)
+        if not certificate_dict:
+            # A secret expired but we did not find matching certificate. Cleaning up
+            event.secret.remove_all_revisions()
+            return
+
+        expiry_time = _get_certificate_expiry_time(certificate_dict["certificate"])
+        if not expiry_time:
+            # A secret expired but matching certificate is invalid. Cleaning up
+            event.secret.remove_all_revisions()
             return
-        self.on.all_certificates_invalidated.emit()
+
+        secret_expiry = event.secret.get_info().expires
+        # Satisfy type checking by ensuring the secret as an expiry time set.
+        # This should always be the case, as the secret just expired.
+        assert secret_expiry is not None
```

Comment:
Why this assert?

Reply:
The comment above explains it. It is only for type checking.